### DISCUSSION TITLE
Defer tasks until after response is sent

### DISCUSSION THREAD

**Discussion by bencroker** [craftcms/cms#15622] [Ideas] [answered]
The [`ApplicationTrait::onAfterRequest()`](https://github.com/craftcms/cms/blob/2ce4b5babe94f6f9b1d016440ea2c8d9817ead34/src/base/ApplicationTrait.php#L505) method is extremely useful and I’ve been wondering about whether it would make sense to add an `ApplicationTrait::onAfterResponse()` method that is triggered after a response is sent to the client. 

The [`Application::STATE_END`](https://www.yiiframework.com/doc/api/2.0/yii-base-application#STATE_END-detail) event could be a candidate for this, or perhaps on a new Craft specific `STATE_AFTER_RESPONSE` event would be favourable. Either way, this new method would ideally execute tasks at the end of both web _and_ console requests.

This would make it possible to defer tasks that should only happen after a response has been sent to the client. Tasks such as garbage collection, email notifications, etc., that should happen in a request in a non-blocking manner. Currently one might reach for queue jobs for such asynchronous tasks, but deferred tasks could be useful for short-lived, non-blocking processes.

**Comment by brandonkelly** [comment, marked as answer]
Tricking PHP into continuing to execute after the response has been sent is pretty hacky. Craft makes it possible via `craft\web\Response::sendAndClose()`, and we call it on queue runner requests before processing the queue. But it’s not something we would want to do automatically on every single request, just in case something has registered an `ON_AFTER_RESPONSE` event handler.

But you’re free to call that `sendAndClose()` method yourself in cases where you need to do post-response work!

**Reply by brandonkelly** [reply]
Worth noting that it’s not 100% reliable, so (like the queue) there should be some way to do the same thing in a non-hacky way for servers that can’t handle `sendAndClose()` properly.

**Reply by bencroker** [reply]
That makes sense, I hadn’t thought about the fact that some servers can’t handle `sendAndClose()`. Thanks for the response!

**Reply by bencroker** [reply]
Just an FYI that a `defer` function was added to Laravel in https://github.com/laravel/framework/pull/52710 with [docs](https://laravel.com/docs/11.x/helpers#deferred-functions). At a quick glance, it looks like it may need some time to become fully stable, so perhaps something to revisit for Craft 6 once the “transition” is further along.
